client/src/controllers/ProgressController.ts

```diff
@@ -0,0 +1,138 @@
+import { Application, Controller } from '@hotwired/stimulus';
+import { gettext } from '../utils/gettext';
+
+/**
+ * <button
+ *   type="submit"
+ *   class="button button-longrunning"
+ *   data-controller="w-progress"
+ *   data-w-progress-active-class="button-longrunning-active"
+ *   data-w-progress-active-value="{% trans 'Signing in…' %}"
+ *   data-w-progress-duration-seconds-value="40"
+ *   data-action="w-progress#activate"
+ * >
+ *  {% icon name="spinner" %}
+ *  <em data-w-progress-target="label">{% trans 'Sign in' %}</em>
+ * </button>
+ */
+export class ProgressController extends Controller {
+  static classes = ['active'];
+  static targets = ['label'];
+  static values = {
+    active: { default: gettext('Loading'), type: String },
+    durationSeconds: { default: 30, type: Number },
+    label: { default: '', type: String },
+    loading: { default: false, type: Boolean },
+  };
+
+  declare activeClass: string;
+  /** Label to use when loading */
+  declare activeValue: string;
+  declare durationSecondsValue: number;
+  /** Label to store the original text on the button */
+  declare labelValue: string;
+  declare loadingValue: boolean;
+  declare readonly hasActiveClass: boolean;
+  declare readonly hasLabelTarget: boolean;
+  declare readonly labelTarget: HTMLElement;
+  timer?: number;
+
+  static afterLoad(identifier: string, application: Application) {
```

Comment:
```suggestion
  /**
   * Ensure we have backwards compatibility with buttons that have
   * not yet adopted the new data attribute syntax.
   * Will warn and advise in release notes that this support
   * will be removed in a future version.
   * @deprecated - RemovedInWagtail60
   */
  static afterLoad(identifier: string, application: Application) {
```

Let's add some context here for our future selves.